itest/rfq_test.go

```diff
@@ -214,6 +215,164 @@ func testRfqAssetBuyHtlcIntercept(t *harnessTest) {
+			// tapd node to send a request for quote message to
+			// Bob's node.
+			PeerPubKey: ts.BobLnd.PubKey[:],
+		},
+	)
+	require.NoError(t.t, err, "unable to upsert asset sell order")
+
+	// Wait until Alice receives an incoming sell quote accept message (sent
+	// from Bob) RFQ event notification.
+	waitErr := wait.NoError(func() error {
+		event, err := aliceEventNtfns.Recv()
+		require.NoError(t.t, err)
+
+		_, ok := event.Event.(*rfqrpc.RfqEvent_PeerAcceptedSellQuote)
+		require.True(t.t, ok, "unexpected event: %v", event)
+
+		return nil
+	}, defaultWaitTimeout)
+	require.NoError(t.t, waitErr)
+
+	// Alice should have received an accepted quote from Bob. This accepted
+	// quote can be used by Alice to make a payment to Bob.
+	acceptedQuotes, err := ts.AliceTapd.QueryPeerAcceptedQuotes(
+		ctxt, &rfqrpc.QueryPeerAcceptedQuotesRequest{},
+	)
+	require.NoError(t.t, err, "unable to query accepted quotes")
+	require.Len(t.t, acceptedQuotes.SellQuotes, 1)
+
+	acceptedQuote := acceptedQuotes.SellQuotes[0]
+
+	// Register to receive RFQ events from Bob's tapd node. We'll use this
+	// to wait for Bob to receive the HTLC with the asset transfer specific
+	// scid.
+	bobEventNtfns, err := ts.BobTapd.SubscribeRfqEventNtfns(
+		ctxb, &rfqrpc.SubscribeRfqEventNtfnsRequest{},
+	)
+	require.NoError(t.t, err)
+
+	// Carol generates and invoice for Alice to settle.
+	addInvoiceResp := ts.CarolLnd.RPC.AddInvoice(&lnrpc.Invoice{
+		ValueMsat: int64(askAmt),
+	})
+	invoice := ts.CarolLnd.RPC.LookupInvoice(addInvoiceResp.RHash)
+
+	// We now need to construct a route for the payment from Alice to Carol.
+	// The route will be Alice -> Bob -> Carol. We'll add the accepted quote
+	// ID as a record to the custom records field of the route's first hop.
+	// This will allow Bob to validate the payment against the accepted
+	// quote.
+	routeBuildRequest := routerrpc.BuildRouteRequest{
+		AmtMsat: int64(askAmt),
+		HopPubkeys: [][]byte{
+			ts.BobLnd.PubKey[:],
+			ts.CarolLnd.PubKey[:],
+		},
+	}
+	routeBuildResp := ts.AliceLnd.RPC.BuildRoute(&routeBuildRequest)
+
+	// Add the accepted quote ID as a record to the custom records field of
+	// the route's first hop.
+	aliceBobHop := routeBuildResp.Route.Hops[0]
+	if aliceBobHop.CustomRecords == nil {
+		aliceBobHop.CustomRecords = make(map[uint64][]byte)
+	}
+
+	aliceBobHop.CustomRecords[rfq.LnCustomRecordType] =
+		acceptedQuote.Id[:]
+
+	// Update the route with the modified first hop.
+	routeBuildResp.Route.Hops[0] = aliceBobHop
+
+	// Send the payment to the route.
+	t.Log("Alice paying invoice")
+	ts.AliceLnd.RPC.SendToRouteV2(&routerrpc.SendToRouteRequest{
+		PaymentHash: invoice.RHash,
+		Route:       routeBuildResp.Route,
+	})
+
+	// At this point Bob should have received a HTLC with the asset transfer
+	// specific scid. We'll wait for Bob to publish an accept HTLC event and
+	// then validate it against the accepted quote.
+	waitErr = wait.NoError(func() error {
```

Comment:
Same here. I think a goroutine that sends on a channel, followed by a `select` on the channel with a timeout is probably what we want.

Reply:
Can't we just block on the `bobEventNtfns.Recv()` call and depend on the itest timeout rather than using `select`? We expect to receive the event.

Reply:
Yes, but then we'd need to initiate `ts.AliceTapd.SubscribeRfqEventNtfns` with a timeout context. Otherwise it will just block forever, since there's nothing to time out (other than eventually `go test` timing out, which currently is set at 20 minutes with `-test.timeout`).

But I think the `BeforeTimeout` function works too!